File: android/app/src/main/java/com/nxtend/team35/yubiboard/diagnostics/AppDiagnostics.kt
package com.nxtend.team35.yubiboard.diagnostics

import android.util.Log
import com.nxtend.team35.yubiboard.BuildConfig
import java.util.ArrayDeque
import java.util.Locale

data class DiagnosticEvent(
    val timestampMs: Long,
    val category: String,
    val name: String,
    val fields: Map<String, String> = emptyMap(),
)

data class DiagnosticSnapshot(
    val events: List<DiagnosticEvent>,
    val counters: Map<String, Long>,
    val gauges: Map<String, String>,
    val metrics: Map<String, MetricSummary>,
)

data class MetricSummary(
    val samples: Int,
    val average: Double,
    val p50: Double,
    val p95: Double,
)

/** Lightweight telemetry controlled by the persisted in-app debug-mode setting. */
object AppDiagnostics {
    private const val LOG_TAG = "YubiBoardDiag"
    private const val MAX_EVENTS = 500
    private const val SAMPLE_INTERVAL_MS = 1_000L
    private val events = ArrayDeque<DiagnosticEvent>()
    private val counters = linkedMapOf<String, Long>()
    private val gauges = linkedMapOf<String, String>()
    private val metrics = linkedMapOf<String, ArrayDeque<Double>>()
    private val lastSamples = mutableMapOf<String, Long>()
    @Volatile
    private var enabled = BuildConfig.DEBUG

    fun setEnabled(value: Boolean) {
        enabled = value
        if (value) event("app", "debug_mode_enabled")
    }

    fun isEnabled(): Boolean = enabled

    @Synchronized
    fun event(category: String, name: String, fields: Map<String, Any?> = emptyMap()) {
        if (!enabled) return
        val normalized = fields.mapValues { (_, value) -> value?.toString().orEmpty() }
        val item = DiagnosticEvent(monotonicMs(), category, name, normalized)
        events.addLast(item)
        while (events.size > MAX_EVENTS) events.removeFirst()
        // android.util.Log is not implemented by local JVM tests.
        runCatching { Log.i(LOG_TAG, item.toJsonLine()) }
    }

    @Synchronized
    fun sampled(key: String, category: String, name: String, fields: Map<String, Any?> = emptyMap()) {
        if (!enabled) return
        val now = monotonicMs()
        val previous = lastSamples[key]
        if (previous != null && now - previous < SAMPLE_INTERVAL_MS) return
        lastSamples[key] = now
        event(category, name, fields)
    }

    @Synchronized
    fun increment(name: String, amount: Long = 1) {
        if (!enabled) return
        counters[name] = (counters[name] ?: 0L) + amount
    }

    @Synchronized
    fun gauge(name: String, value: Any?) {
        if (!enabled) return
        gauges[name] = value?.toString().orEmpty()
    }

    @Synchronized
    fun metric(name: String, value: Number) {
        if (!enabled) return
        val values = metrics.getOrPut(name) { ArrayDeque() }
        values.addLast(value.toDouble())
        while (values.size > MAX_METRIC_SAMPLES) values.removeFirst()
    }

    @Synchronized
    fun snapshot(): DiagnosticSnapshot = DiagnosticSnapshot(
        events = events.toList(),
        counters = counters.toMap(),
        gauges = gauges.toMap(),
        metrics = metrics.mapValues { (_, values) -> values.summarize() },
    )

    @Synchronized
    fun clear() {
        events.clear()
        counters.clear()
        gauges.clear()
        metrics.clear()
        lastSamples.clear()
    }

    fun format(snapshot: DiagnosticSnapshot = snapshot()): String = buildString {
        appendLine("YubiBoard debug diagnostics")
        if (snapshot.gauges.isNotEmpty()) {
            appendLine("\nLatest values")
            snapshot.gauges.forEach { (key, value) -> appendLine("$key: $value") }
        }
        if (snapshot.counters.isNotEmpty()) {
            appendLine("\nCounters")
            snapshot.counters.forEach { (key, value) -> appendLine("$key: $value") }
        }
        if (snapshot.metrics.isNotEmpty()) {
            appendLine("\nRolling metrics")
            snapshot.metrics.forEach { (key, value) ->
                appendLine(
                    "$key: n=${value.samples} avg=${"%.2f".format(Locale.US, value.average)} " +
                        "p50=${"%.2f".format(Locale.US, value.p50)} p95=${"%.2f".format(Locale.US, value.p95)}",
                )
            }
        }
        appendLine("\nRecent events")
        snapshot.events.takeLast(80).forEach { event ->
            append("${event.timestampMs} ${event.category}/${event.name}")
            if (event.fields.isNotEmpty()) {
                append(" ")
                append(event.fields.entries.joinToString { "${it.key}=${it.value}" })
            }
            appendLine()
        }
    }

    fun jsonLines(snapshot: DiagnosticSnapshot = snapshot()): String =
        snapshot.events.joinToString(separator = "\n", postfix = if (snapshot.events.isEmpty()) "" else "\n") {
            it.toJsonLine()
        }

    private fun DiagnosticEvent.toJsonLine(): String = buildString {
        append("{\"timestampMs\":")
        append(timestampMs)
        append(",\"category\":\"")
        append(category.jsonEscape())
        append("\",\"name\":\"")
        append(name.jsonEscape())
        append("\",\"fields\":{")
        append(fields.entries.joinToString { (key, value) ->
            "\"${key.jsonEscape()}\":\"${value.jsonEscape()}\""
        })
        append("}}")
    }

    private fun String.jsonEscape(): String = buildString(length) {
        this@jsonEscape.forEach { character ->
            when (character) {
                '\\' -> append("\\\\")
                '"' -> append("\\\"")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                else -> if (character.code < 0x20) {
                    append(String.format(Locale.US, "\\u%04x", character.code))
                } else {
                    append(character)
                }
            }
        }
    }

    private fun monotonicMs(): Long = System.nanoTime() / 1_000_000L

    private fun Collection<Double>.summarize(): MetricSummary {
        val sorted = sorted()
        fun percentile(fraction: Double): Double {
            if (sorted.isEmpty()) return 0.0
            val index = (sorted.size * fraction).toInt().coerceIn(sorted.indices)
            return sorted[index]
        }
        return MetricSummary(
            samples = sorted.size,
            average = if (sorted.isEmpty()) 0.0 else sorted.average(),
            p50 = percentile(0.50),
            p95 = percentile(0.95),
        )
    }

    private const val MAX_METRIC_SAMPLES = 300
}

File: ios/Screact/Diagnostics/AppDiagnostics.swift
import Foundation
import os

/// Lightweight telemetry, gated by the in-app debug-mode setting. Port of Android's AppDiagnostics
/// (trimmed: keeps events/counters/gauges/metrics/sampled and the text/JSONL formatters).
struct DiagnosticEvent {
    let timestampMs: Int64
    let category: String
    let name: String
    let fields: [String: String]
}

struct MetricSummary {
    let samples: Int
    let average: Double
    let p50: Double
    let p95: Double
}

struct DiagnosticSnapshot {
    let events: [DiagnosticEvent]
    let counters: [String: Int64]
    let gauges: [String: String]
    let metrics: [String: MetricSummary]
}

final class AppDiagnostics {
    static let shared = AppDiagnostics()

    private let queue = DispatchQueue(label: "com.nxtend.team35.yubiboard.diagnostics")
    private let logger = Logger(subsystem: "com.nxtend.team35.yubiboard", category: "YubiBoardDiag")
    private let maxEvents = 500
    private let maxMetricSamples = 300
    private let sampleIntervalMs: Int64 = 1_000

    private var events: [DiagnosticEvent] = []
    private var counters: [String: Int64] = [:]
    private var gauges: [String: String] = [:]
    private var metrics: [String: [Double]] = [:]
    private var lastSamples: [String: Int64] = [:]
    private var enabled = false

    func setEnabled(_ value: Bool) {
        queue.sync { enabled = value }
        if value { event("app", "debug_mode_enabled") }
    }

    var isEnabled: Bool { queue.sync { enabled } }

    func event(_ category: String, _ name: String, _ fields: [String: Any?] = [:]) {
        queue.sync {
            guard enabled else { return }
            let normalized = fields.mapValues { value -> String in
                guard let value = value else { return "" }
                return String(describing: value)
            }
            let item = DiagnosticEvent(timestampMs: Self.monotonicMs(), category: category, name: name, fields: normalized)
            events.append(item)
            if events.count > maxEvents { events.removeFirst(events.count - maxEvents) }
            logger.info("\(category)/\(name)")
        }
    }

    func sampled(_ key: String, _ category: String, _ name: String, _ fields: [String: Any?] = [:]) {
        let shouldEmit: Bool = queue.sync {
            guard enabled else { return false }
            let now = Self.monotonicMs()
            if let previous = lastSamples[key], now - previous < sampleIntervalMs { return false }
            lastSamples[key] = now
            return true
        }
        if shouldEmit { event(category, name, fields) }
    }

    func increment(_ name: String, _ amount: Int64 = 1) {
        queue.sync {
            guard enabled else { return }
            counters[name, default: 0] += amount
        }
    }

    func gauge(_ name: String, _ value: Any?) {
        queue.sync {
            guard enabled else { return }
            gauges[name] = value.map { String(describing: $0) } ?? ""
        }
    }

    func metric(_ name: String, _ value: Double) {
        queue.sync {
            guard enabled else { return }
            metrics[name, default: []].append(value)
            if metrics[name]!.count > maxMetricSamples {
                metrics[name]!.removeFirst(metrics[name]!.count - maxMetricSamples)
            }
        }
    }

    func snapshot() -> DiagnosticSnapshot {
        queue.sync {
            DiagnosticSnapshot(
                events: events,
                counters: counters,
                gauges: gauges,
                metrics: metrics.mapValues { Self.summarize($0) }
            )
        }
    }

    func clear() {
        queue.sync {
            events.removeAll()
            counters.removeAll()
            gauges.removeAll()
            metrics.removeAll()
            lastSamples.removeAll()
        }
    }

    private static func summarize(_ values: [Double]) -> MetricSummary {
        let sorted = values.sorted()
        func percentile(_ fraction: Double) -> Double {
            guard !sorted.isEmpty else { return 0 }
            let index = min(max(Int(Double(sorted.count) * fraction), 0), sorted.count - 1)
            return sorted[index]
        }
        return MetricSummary(
            samples: sorted.count,
            average: sorted.isEmpty ? 0 : sorted.reduce(0, +) / Double(sorted.count),
            p50: percentile(0.50),
            p95: percentile(0.95)
        )
    }

    static func monotonicMs() -> Int64 {
        Int64(DispatchTime.now().uptimeNanoseconds / 1_000_000)
    }
}
